.github/workflows/pull-request-with-code.yml

```diff
@@ -17,35 +17,62 @@ on:
       - '**/*.toml'
   workflow_dispatch:
 
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: ${{ github.ref != 'refs/heads/master' }}
+
 env:
-  ORG_GRADLE_PROJECT_signingKey : ${{ secrets.ORG_GRADLE_PROJECT_SIGNINGKEY }}
+  ORG_GRADLE_PROJECT_signingKey: ${{ secrets.ORG_GRADLE_PROJECT_SIGNINGKEY }}
   ORG_GRADLE_PROJECT_signingKeyId: ${{ secrets.ORG_GRADLE_PROJECT_SIGNINGKEYID }}
-  ORG_GRADLE_PROJECT_signingKeyPassword : ${{ secrets.ORG_GRADLE_PROJECT_SIGNINGPASSWORD }}
+  ORG_GRADLE_PROJECT_signingKeyPassword: ${{ secrets.ORG_GRADLE_PROJECT_SIGNINGPASSWORD }}
   CLI_TEST_MAX_DURATION_IN_SECONDS: 10
 
 # Note that "jobs.build.strategy" and "jobs.build.runs-on" should be kept in sync with "pull-request-wth-code"
 jobs:
   build:
     strategy:
       matrix:
-        os: [ ubuntu-latest, windows-latest ]
-        # When changing the list of JDK versions, the build configuration has to be changed by a repository admin. See
-        # https://github.com/pinterest/ktlint/pull/1787#issuecomment-1409074092
-        jdk: [ 8, 11, 17, 19 ]
-        exclude: # windows with JDK8 are *really* flaky
-          - os: windows-latest
-            jdk: 8
+        os: [ubuntu-latest, windows-latest]
     runs-on: ${{ matrix.os }}
+    name: "[build] OS: ${{ matrix.os }}"
     steps:
       - uses: actions/checkout@v3
+
       - uses: actions/setup-java@v3
         with:
           distribution: 'zulu'
-          java-version: ${{ matrix.jdk }}
+          java-version: 20 # This should match all `jvmToolchain(xxx)` calls in the project
+
       - uses: gradle/gradle-build-action@v2
         with:
           gradle-home-cache-cleanup: true
+
       - name: Build with release Kotlin version
         run: ./gradlew build ktlintCheck --no-configuration-cache
-      - name: Build with dev Kotlin version
+
+      - name: Build with assemble Kotlin version
         run: ./gradlew -PkotlinDev build ktlintCheck --no-configuration-cache
+
+  tests:
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [ubuntu-latest, windows-latest]
+        jdk: [ 8, 11, 17 ] # list of Java versions to run tests against
+    runs-on: ${{ matrix.os }}
+    name: "[tests] OS: ${{ matrix.os }}, Java: ${{ matrix.jdk }}"
+    steps:
+      - uses: actions/checkout@v3
+
+      - uses: actions/setup-java@v3
+        with:
+          distribution: 'zulu'
+          java-version: |
+            ${{ matrix.jdk }}
+            20
```

Comment:
Add ```# This should match all `jvmToolchain(xxx)` calls in the project```?

Reply:
I couldn't add the line directly next to `20`, so I added it to the closest place possible (and I hope it'll pass the CI).
Heads up: I have already another improvement to the CI queued up ([preview](https://github.com/mateuszkwiecinski/ktlint/compare/set_toolchains...mateuszkwiecinski:ktlint:toolchains_testing)), where I'll try to unify how workflows set up the environment for Gradle builds (so I don't have to hardcode <current> java version that many times).